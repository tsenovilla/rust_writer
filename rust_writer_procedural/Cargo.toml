[package]
name = "rust_writer_procedural"
version.workspace = true
repository.workspace = true 
license.workspace = true
rust-version.workspace = true
edition.workspace = true
description.workspace = true
documentation.workspace = true

[dependencies]
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true, features = ["full", "extra-traits", "parsing"] }

[lib]
proc-macro = true

[dev-dependencies]
trybuild = "1.0.103"
rust_writer = { workspace = true }
test_builder = { workspace = true }
