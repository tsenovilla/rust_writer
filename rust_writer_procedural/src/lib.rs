// SPDX-License-Identifier: GPL-3.0

mod finder;
pub(crate) mod helpers;
mod mutator;
pub(crate) mod parse;

use proc_macro::TokenStream;

/// # Description
///
/// The mutator macro is used to define a new implementor which combines a  
///
/// # Compatibility with [`#[finder]`](https://docs.rs/rust_writer/latest/rust_writer/ast/attr.finder.html)
#[proc_macro_attribute]
pub fn mutator(attrs: TokenStream, item: TokenStream) -> TokenStream {
	mutator::mutator(attrs, item)
}

#[proc_macro_attribute]
pub fn local_mutator(_: TokenStream, item: TokenStream) -> TokenStream {
	mutator::local_mutator(item)
}

#[proc_macro_attribute]
pub fn finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::finder(attrs, item)
}

#[proc_macro_attribute]
pub fn local_finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::local_finder(attrs, item)
}

#[doc(hidden)]
#[proc_macro_attribute]
pub fn already_expanded(_: TokenStream, item: TokenStream) -> TokenStream {
	item
}
