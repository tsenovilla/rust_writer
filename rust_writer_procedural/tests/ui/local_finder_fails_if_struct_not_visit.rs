// SPDX-License-Identifier: GPL-3.0

use rust_writer_procedural::local_finder;

#[local_finder('a)]
#[derive(Debug)]
struct SomeStruct {
	found: [bool; 1],
}

fn main() {}
