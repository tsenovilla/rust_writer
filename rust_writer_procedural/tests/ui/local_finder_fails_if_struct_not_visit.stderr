error[E0599]: no method named `visit_file` found for mutable reference `&mut SomeStruct` in the current scope
 --> tests/ui/local_finder_fails_if_struct_not_visit.rs:5:1
  |
5 | #[local_finder('a)]
  | ^^^^^^^^^^^^^^^^^^^ method not found in `&mut SomeStruct`
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `visit_file`, perhaps you need to implement it:
          candidate #1: `Visit`
  = note: this error originates in the attribute macro `local_finder` (in Nightly builds, run with -Z macro-backtrace for more info)
